handler/src/main/java/io/netty/handler/ssl/SslContext.java

```diff
@@ -673,6 +673,204 @@ static ApplicationProtocolConfig toApplicationProtocolConfig(Iterable<String> ne
         return apn;
     }
 
+    /**
+     * Creates a builder for new client-side {@link SslContext}.
+     */
+    public static Builder newBuilderForClient() {
+        return new Builder(true);
+    }
+
+    /**
+     * Creates a builder for new server-side {@link SslContext}.
+     *
+     * @param keyCertChainFile an X.509 certificate chain file in PEM format
+     * @param keyFile a PKCS#8 private key file in PEM format
+     */
+    public static Builder newBuilderForServer(File keyCertChainFile, File keyFile) {
+        return new Builder(false).keyManager(keyCertChainFile, keyFile);
+    }
+
+    /**
+     * Creates a builder for new server-side {@link SslContext}.
+     *
+     * @param keyCertChainFile an X.509 certificate chain file in PEM format
+     * @param keyFile a PKCS#8 private key file in PEM format
+     * @param keyPassword the password of the {@code keyFile}, or {@code null} if it's not
+     *     password-protected
+     */
+    public static Builder newBuilderForServer(File keyCertChainFile, File keyFile,
+            String keyPassword) {
+        return new Builder(false).keyManager(keyCertChainFile, keyFile, keyPassword);
+    }
+
+    /** Builder for configuring a new SslContext for creation. */
+    public static final class Builder {
```

Comment:
> This static factory pattern lead to us wanting to create the builder.

I think that is throwing out the baby with the bath water. The extra methods make the API much more obvious to use and gives clear errors.

I feel like your argument applies equally to keyManager() itself. But this sucks as it doesn't guide the user at all and figuring out the interaction between the various methods is complex/strewn about:
```java
public Builder keyCertChainFile(File keyCertChainFile)
public Builder keyFile(File keyFile)
public Builder keyPassword(String keyPassword)
```

> For example why don't we have client versions of these

Because they aren't required. Both client and server can use keyManager(); it is only required for the server though. The point isn't to provide conveniences for every way it may need to be constructed, but instead to make it easy to keep the Builder in a consistent state. Having simply keyManager() vs keyCertChainFile(), keyFile() and keyPassword() contributes to the same objective.

> then why not trust manager parameters or key manager parameters

Key manager is there, but trust manager isn't because it is optional. There are very few required options and by API stability restrictions there can't be additional ones in the future.